### ISSUE TITLE
Followup from #83174

### ISSUE THREAD

**Issue by camelid** [rust-lang/rust#90425]
Things I'm planning to follow up on once #83174 is merged:

- https://github.com/rust-lang/rust/pull/83174#discussion_r733079602

**Comment by camelid** [comment]
Case of a confusing span:

```rust
fn main() {
    let mut v = vec![""];
    v.push({
        let x = v.pop().unwrap();
        x
    });
}
```

```rust
error[E0499]: cannot borrow `v` as mutable more than once at a time
 --> src/main.rs:4:17
  |
3 |       v.push({
  |       - ---- first borrow later used by call
  |  _____|
  | |
4 | |         let x = v.pop().unwrap();
  | |                 ^^^^^^^ second mutable borrow occurs here
5 | |         x
6 | |     });
  | |______- first mutable borrow occurs here
  |
help: try adding a local storing this argument...
 --> src/main.rs:4:17
  |
4 |         let x = v.pop().unwrap();
  |                 ^^^^^^^
help: ...and then using that local as the argument to this call
 --> src/main.rs:3:5
  |
3 | /     v.push({
4 | |         let x = v.pop().unwrap();
5 | |         x
6 | |     });
  | |______^
```

**Comment by camelid** [comment]
@Trolldemorted So as far as I understand it, this behavior is correct according to the way the Rust language is defined (not just the implementation). The issue is that this code turns into something like

```rust
let self1 = &mut self;             // self1: lifetime starts here
let self2 = &mut self;             // self2: lifetime starts here
let bar_result = Self::bar(self2); // self2: lifetime ends here
Self::foo(arg1, bar_result);       // self1: lifetime ends here
```

But you're not allowed to have two mutable borrows active at the same time; they're supposed to be unique.

**Comment by Trolldemorted** [comment]
Fair point, so a backwards-compatible fix for this bug would involve two slightly different SSAs, one for the borrow checker, one for the compiler?

Two-phase-borrows also disrupt the normal SSA flow, don't they?

**Comment by camelid** [comment]
> Fair point, so a backwards-compatible fix for this bug would involve two slightly different SSAs, one for the borrow checker, one for the compiler?
> 
> Two-phase-borrows also disrupt the normal SSA flow, don't they?

@Trolldemorted apologies for never replying to your message. Regarding your questions:

1. Using a meaningfully different version of the MIR for the borrow checker seems quite undesirable to me. Currently, the borrow checker runs on the same MIR that is then desugared a bit more and passed to the code generator. If we changed the actual order of statements within the MIR for borrow checking, it would add significant complexity and mean that the checking is done on a fictional version of the program. This opens many opportunities for bugs and also unintuitive behavior.

2. Two-phase borrows do _not_ change the order of evaluation in the program. They simply allow _shared_ borrows to be taken out and used in between when a _mutable_ borrow is taken out and when it is used. This is okay because the shared borrows cannot invalidate the address of the data, so the mutable borrow's address-value will still be correct when it is eventually used.

Two-phase borrows _could_ in theory be extended to allow mutable borrows in between creation and activation of a shared borrow, or mutable borrows in between creation and activation of another mutable borrow (like the case we were discussing). This idea was even [raised](https://rust-lang.github.io/rfcs/2025-nested-method-calls.html#permit-more-things-during-the-restricted-period) as a possibility in the RFC for two-phase borrows:

> The current notion of a ‘restricted’ borrow is identical to a shared borrow. However, we could in principle permit more things during the restricted period – basically we could permit anything that does not invalidate the reference we created.
>
> [...]
>
> We opted against this variation for several reasons:
>
> - It makes the borrow checker more complex by introducing not only two-phase borrows, but a new set of restrictions that must be worked out in detail. The current RFC leverages the existing category of shared borrows.
>
> - The main gain here is the ability to intersperse two mutable calls (as in the example), or to have an outer shared borrow with an inner mutable borrow. In general, this implies that there is some careful ordering of mutation going on here: in particular, the outer method call will observe the state changes made by the inner calls. This feels like a case where it is helpful to have the user pull the two calls apart, so that their relative side-effects are clearly visible.
>
>Of course, it would be possible to loosen the rules in the future.

**Comment by Enselic** [comment]
Triage: Hmm, I don't understand what this issue is about. What needs to be done before we can close this issue?

**Comment by Enselic** [comment]
Seems like no one else does either :) Closing as Obsolete. If there is still work to be done, I recommend opening new issues with fresh descriptions.

**Comment by camelid** [comment]
Sorry, didn't see your triage comment. This issue was mostly a self-reminder to improve some outstanding issues with the diagnostic. But it's fine to close to clean up the issue queue. 👍
